File: leetcode-algorithms/003. Longest Substring Without Repeating Characters/answer.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        v_i = {}
        j = 0
        res = 0
        for i,v in enumerate(s):
        	if v in v_i:
        		j = max(j,v_i[v]+1)
        	v_i[v] = i
        	res = max(res,i-j+1)
        return res




if __name__ == '__main__':
	s = Solution()
	ss = 'abcababcdefg'
	res = s.lengthOfLongestSubstring(ss)
	print(res)

File: leetcode-algorithms/003. Longest Substring Without Repeating Characters/answer.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <map>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res = 0;
        map<char,int> m;
        for(auto i=0,j=0; i<s.size(); i++){
        	if (m.find(s[i]) != m.end())
        		j = max(j,m[s[i]]+1);
        	m[s[i]] = i;
        	res = max(res,i-j+1);
        }
        return res;
    }
};

int main()
{
	string s("abcdabcde");
	Solution solution;
	int res = solution.lengthOfLongestSubstring(s);
	cout<<res<<endl;
	return 0;
}